src/py/flwr/client/grpc_rere_client/grpc_adapter.py

```diff
@@ -0,0 +1,126 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Grpc Adapter."""
+
+
+import sys
+from logging import DEBUG
+from typing import Any, Type, TypeVar, cast
+
+from google.protobuf.message import Message as GrpcMessage
+
+from flwr.common import log
+from flwr.common.constant import (
+    GRPC_ADAPTER_METADATA_FLOWER_VERSION_KEY,
+    GRPC_ADAPTER_METADATA_SHOULD_EXIT_KEY,
+)
+from flwr.common.version import package_version
+from flwr.proto.fleet_pb2 import (  # pylint: disable=E0611
+    CreateNodeRequest,
+    CreateNodeResponse,
+    DeleteNodeRequest,
+    DeleteNodeResponse,
+    PingRequest,
+    PingResponse,
+    PullTaskInsRequest,
+    PullTaskInsResponse,
+    PushTaskResRequest,
+    PushTaskResResponse,
+)
+from flwr.proto.grpcadapter_pb2 import MessageContainer  # pylint: disable=E0611
+from flwr.proto.grpcadapter_pb2_grpc import GrpcAdapterStub
+from flwr.proto.run_pb2 import GetRunRequest, GetRunResponse  # pylint: disable=E0611
+
+T = TypeVar("T", bound=GrpcMessage)
+
+
+class GrpcAdapter:
+    """The Adapter class to send and receive gRPC messages via the GrpcAdapterStub.
+
+    This class utilizes the GrpcAdapterStub to send and receive gRPC messages which are
+    defined and used by the Fleet API, as defined in `fleet.proto`.
+    """
+
+    def __init__(self, channel: Any) -> None:
+        self.stub = GrpcAdapterStub(channel)
+
+    def _send_and_receive(
+        self, request: GrpcMessage, response_type: Type[T], **kwargs: Any
+    ) -> T:
+        # Serialize request
+        container_req = MessageContainer(
+            metadata={GRPC_ADAPTER_METADATA_FLOWER_VERSION_KEY: package_version},
+            grpc_message_name=request.__class__.__qualname__,
+            grpc_message_content=request.SerializeToString(),
+        )
+
+        # Send via the stub
+        container_res = cast(
+            MessageContainer, self.stub.SendReceive(container_req, **kwargs)
+        )
+
+        # Handle control message
+        should_exit = (
+            container_res.metadata.get(GRPC_ADAPTER_METADATA_SHOULD_EXIT_KEY, "false")
+            == "true"
+        )
+        if should_exit:
+            log(DEBUG, "Received shutdown signal: exit flag is set to True. Exiting...")
+            sys.exit(0)
+
+        # Check the grpc_message_name of the response
+        if container_res.grpc_message_name != response_type.__qualname__:
+            raise ValueError(
+                f"Invalid grpc_message_name. Expected {response_type.__qualname__}"
+                f", but got {container_res.grpc_message_name}."
+            )
+
+        # Deserialize response
+        response = response_type()
+        response.ParseFromString(container_res.grpc_message_content)
+        return response
+
+    # pylint: disable=C0103
```

Comment:
Should we add this to every line? Right now it's also disabled for the implementation of each method, but it's only required for each method name